src/mpi/init/init_async.c

```diff
@@ -168,6 +168,8 @@ int MPIR_Finalize_async_thread(void)
     MPID_Thread_mutex_unlock(&progress_mutex, &mpi_errno);
     MPIR_Assert(!mpi_errno);
 
+    MPID_Thread_join(progress_thread_id);
```

Comment:
Is the above `MPID_Thread_cond_wait` still necessary with this join synchronization?

Reply:
The `cond_wait` is the first synchronization. It put the main thread to sleep while the other threads finish the progress.

Reply:
I'm still confused. Before the init refactoring, cond wait was the only synchronization we did with the progress thread, and we were free of deadlocks in finalize. Now we have to synchronize twice (actually 3 times if you count send/recv). I am trying to understand why that is.

Reply:
From our offline discussion, I suspect the old cond_wait code was essentially an implicit thread join operation. However now we are explicit, maybe it is not needed? In practice it will deadlock in my experiments...